[E2E] enhance pod polling failures

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This makes pod timeout failures easier to read (less clutter) and more informative (last pod state is part of failure message).

It also adds tests for this particular framework behavior.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/cc @tallclair @aojea

pohly committed:
e2e: include more information in pod timeout error

The advantage is that the extra error information is guaranteed to be printed
directly before the failure and we avoid one extra log line that would have to
be correlated with the failure.

pohly commented:
It's ironic that a PR about polling runs into this (unrelated) flake (https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/109828/pull-kubernetes-e2e-gce-ubuntu-containerd/1522227643875856384/):
```
Kubernetes  e2e suite: [sig-api-machinery] ResourceQuota should create a  ResourceQuota and capture the life of a configMap. [Conformance]

Failure test/e2e/framework/framework.go:647
May  5 15:36:23.511: Unexpected error:
    <*errors.errorString | 0xc0001dc2b0>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
occurred
test/e2e/apimachinery/resource_quota.go:360}
```

pohly commented:
Another failure is also instructive (https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/109828/pull-kubernetes-conformance-kind-ipv6-parallel/1522227643787776000):
```
May  5 15:32:36.308: Unexpected error:
    <*errors.errorString | 0xc002264960>: {
        s: "expected pod \"pod-4cb63e74-6f21-4690-8ba2-3e5b490df0f0\" success: timed out while waiting for pod ...
```

Here the simplified failure printing isn't done because the error isn't getting wrapped properly:
https://github.com/kubernetes/kubernetes/blob/751e9c0f5b8af2c9f0e5f63e019c8e9fa483306f/test/e2e/framework/util.go#L873

Should we do a search/replace of `%v .*err` with `%w`? How do we ensure that this gets used consistently going forward?

pohly commented:
Another sub-optimal timeout failure (https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/109828/pull-kubernetes-conformance-kind-ga-only-parallel/1522304400956067840):

```
{ Failure test/e2e/framework/framework.go:647
May  5 20:21:18.772: wait for pod "alpine-nnp-false-3642e7a4-ed94-40e8-ac29-a5c624201bdf" to succeed
Expected success, but got an error:
    <*fmt.wrapError | 0xc002c2cd40>: {
        msg: "error while waiting for pod security-context-test-1238/alpine-nnp-false-3642e7a4-ed94-40e8-ac29-a5c624201bdf to be Succeeded or Failed: pod \"alpine-nnp-false-3642e7a4-ed94-40e8-ac29-a5c624201bdf\" failed with reason: \"\", message: \"\"",
        err: {
            s: "pod \"alpine-nnp-false-3642e7a4-ed94-40e8-ac29-a5c624201bdf\" failed with reason: \"\", message: \"\"",
        },
    }
    error while waiting for pod security-context-test-1238/alpine-nnp-false-3642e7a4-ed94-40e8-ac29-a5c624201bdf to be Succeeded or Failed: pod "alpine-nnp-false-3642e7a4-ed94-40e8-ac29-a5c624201bdf" failed with reason: "", message: ""
test/e2e/framework/pods.go:240}
```

The problem in test/e2e/framework/pods.go is two-fold:
- the failure is attributed to the helper function (pods.go:240) instead of the caller
- gomega.Succeed is used to check the error instead of ExpectNoError, thus bypassing the enhanced failure printing in this PR

Does anybody remember what the purpose of these Gomega wrapper functions are? Are they meant to be syntactic sugar and thus optional or are they part of some plan to get rid of Gomega, i.e. should be used everywhere?

Now an argument could be made for requiring the use of the framework wrapper functions (better failure message handling).

@Random-Liu: you have one TODO on PodClient.WaitForSuccess, the function that failed here. Any comments?

aojea commented:
> Should we do a search/replace of `%v .*err` with `%w`? How do we ensure that this gets used consistently going forward?

Thockin tried with https://github.com/quasilyte/go-ruleguard but it needs to import an external module just for linting :shrug:

pohly commented:
If we limit that dependency to `hack/tools` it might be fine.

But first we need to decide what the rule should be - always wrap the underlying error? That is not necessarily always the right approach, at least not outside of test/e2e.

tallclair commented on test/e2e/framework/pods.go:
```diff
@@ -90,24 +102,35 @@ func (f *Framework) PodClientNS(namespace string) *PodClient {
+}
+
+func (c *PodClient) createSync(skipFrames int, pod *v1.Pod) *v1.Pod {
+	skipFrames++
```
This `skipFrames` stuff is a bit hard to follow, and seems fragile. What if instead you just iterate through the callers until you get one outside this file? e.g.

```go
func (c *PodClient) callerOffset() int {
  _, podClientFile, _, ok := runtime.Caller(0)
  if !ok {
    return c.skipFrames+2; // callerOffset() + caller
  }
  for i := 2; i < MaxDepth; i++ {
    _, f, _, ok := runtime.Caller(i)
    if f != podClientFile || !ok {
      return c.skipFrames + i
    }
  }
  // call stack too deep
  return c.skipFrames + MaxDepth
}
```

pohly replied:
That doesn't work when the code in pods.go is called by some other helper function. Then we would log that helper function instead of the test code.

Skipping frames by counting them is a normal pattern for logging and also how Gomega does it. I agree that off-by-one errors are problematic, but I don't see a good alternative. The "testing" package allows helper functions to mark themselves as helper, but that only works in unit tests and for stack unwinding by the standard library.

tallclair commented on test/e2e/framework/pods.go:
```diff
@@ -301,6 +326,6 @@ func (c *PodClient) MatchContainerOutput(name string, containerName string, expe
 func (c *PodClient) PodIsReady(name string) bool {
 	pod, err := c.Get(context.TODO(), name, metav1.GetOptions{})
-	ExpectNoError(err)
+	ExpectNoErrorWithOffset(c.skipFrames+1, err, "get pod")
```
```suggestion
	ExpectNoErrorWithOffset(c.skipFrames+1, err, "getting pod to check PodIsReady")
```

pohly commented on test/e2e/framework-tests/framework-tests.go:
```diff
@@ -0,0 +1,106 @@
+	return ginkgo.Describe("[sig-testing] "+text, body)
+}
+
+var _ = SIGDescribe("E2E framework", func() {
```
I'll try that approach.

pohly replied:
Hmm, are those unit tests supposed to pass?

```
$ go test ./test/e2e/framework
Running Suite: Logging Suite
============================
Random Seed: 1652257205
Will run 4 of 4 specs

• Failure [0.002 seconds]
log
/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/log_test.go:50
  fails [It]
  /nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/log_test.go:54

  May 11 10:20:05.325: I'm failing.

  /nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/log_test.go:57
...
Ran 4 of 4 Specs in 0.010 seconds
FAIL! -- 0 Passed | 4 Failed | 0 Pending | 0 Skipped
--- FAIL: TestFailureOutput (0.02s)
panic: 
Your test failed.
Ginkgo panics to prevent subsequent assertions from running.
Normally Ginkgo rescues this panic so you shouldn't see it.

But, if you make an assertion in a goroutine, Ginkgo can't capture the panic.
To circumvent this, you should call

	defer GinkgoRecover()

at the top of the goroutine that caused this panic.
 [recovered]
	panic: 
Your test failed.
Ginkgo panics to prevent subsequent assertions from running.
Normally Ginkgo rescues this panic so you shouldn't see it.

But, if you make an assertion in a goroutine, Ginkgo can't capture the panic.
To circumvent this, you should call

	defer GinkgoRecover()

at the top of the goroutine that caused this panic.
 [recovered]
	panic: 
Your test failed.
Ginkgo panics to prevent subsequent assertions from running.
Normally Ginkgo rescues this panic so you shouldn't see it.

But, if you make an assertion in a goroutine, Ginkgo can't capture the panic.
To circumvent this, you should call

	defer GinkgoRecover()

at the top of the goroutine that caused this panic.


goroutine 130 [running]:
testing.tRunner.func1.2({0x19cca20, 0xc000915680})
	/nvme/gopath/go-1.18.1/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/nvme/gopath/go-1.18.1/src/testing/testing.go:1392 +0x39f
panic({0x19cca20, 0xc000915680})
	/nvme/gopath/go-1.18.1/src/runtime/panic.go:838 +0x207
k8s.io/kubernetes/test/e2e/framework/ginkgowrapper.Fail.func1()
	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/ginkgowrapper/wrapper.go:63 +0x7d
panic({0x17ca9e0, 0x1d8a140})
	/nvme/gopath/go-1.18.1/src/runtime/panic.go:838 +0x207
github.com/onsi/ginkgo.Fail({0xc0007c26c0, 0x8e}, {0xc0000298f0?, 0xc000029900?, 0x0?})
	/nvme/gopath/pkg/mod/github.com/onsi/ginkgo@v1.14.0/ginkgo_dsl.go:260 +0xdd
k8s.io/kubernetes/test/e2e/framework/ginkgowrapper.Fail({0xc0007c26c0, 0x8e}, {0xc0000299d0?, 0x1afce88?, 0xc0000299f0?})
	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/ginkgowrapper/wrapper.go:67 +0x197
k8s.io/kubernetes/test/e2e/framework.Fail({0xc0008ca700, 0x79}, {0xc0008b9a30?, 0xc00091da40?, 0xc000029a90?})
	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/log.go:63 +0x145
github.com/onsi/gomega/internal/assertion.(*Assertion).match(0xc000029b40, {0x1d99e20, 0xc0008e4310}, 0x1, {0xc000029cb8, 0x2, 0x2})
	/nvme/gopath/pkg/mod/github.com/onsi/gomega@v1.10.1/internal/assertion/assertion.go:79 +0x1bd
github.com/onsi/gomega/internal/assertion.(*Assertion).To(0xc000029b40, {0x1d99e20, 0xc0008e4310}, {0xc000029cb8, 0x2, 0x2})
	/nvme/gopath/pkg/mod/github.com/onsi/gomega@v1.10.1/internal/assertion/assertion.go:38 +0x95
k8s.io/kubernetes/test/e2e/framework.ExpectEqual({0x17ca9e0?, 0xc0008e42f0?}, {0x17ca9e0?, 0xc0008e4300}, {0xc000029cb8, 0x2, 0x2})
	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/expect.go:29 +0x125
k8s.io/kubernetes/test/e2e/framework_test.TestFailureOutput(0x0?)
	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/log_test.go:123 +0x77b
testing.tRunner(0xc000600d00, 0x1c0d208)
	/nvme/gopath/go-1.18.1/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/nvme/gopath/go-1.18.1/src/testing/testing.go:1486 +0x35f
FAIL	k8s.io/kubernetes/test/e2e/framework	0.070s
FAIL
```

Those tests are not run by "make test". Not sure whether that is intentional?

https://github.com/kubernetes/kubernetes/blob/b74d023e70d6064c7f3f77031e7d26ec38497fc9/hack/make-rules/test.sh#L47

aojea replied:
I can't see the goroutine that panics :thinking:

pohly replied:
It's the goroutine which executes TestFailureOutput. "go test" internally uses multiple different goroutines.

pohly replied:
Fix is in https://github.com/kubernetes/kubernetes/pull/109961

pohly replied:
> I'll try that approach.

It works. PR #109961 must be merged first, then I can follow with tests for the current e2epod code. Here's a preview: https://github.com/kubernetes/kubernetes/compare/master...pohly:pod-wait-tests?expand=1

@tallclair: can you help get that PR merged?

pohly replied:
I just realized something... I'm the one who wrote this `log_test.go`. I had completely forgotten about that. :sweat_smile:

pohly commented on test/e2e/framework-tests/framework-tests.go:
```diff
@@ -0,0 +1,106 @@
+
+var _ = SIGDescribe("E2E framework", func() {
+	f := framework.NewDefaultFramework("framework-test")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelBaseline
```
I tried for a while, but pod creation always failed with permission errors. I know, I should try harder :sweat_smile:

tallclair replied:
The utilities added in https://github.com/kubernetes/kubernetes/pull/109946 should make this easier. Looks like it's no longer relevant here though.

pohly commented on test/e2e/framework/e2eerrors/errors.go:
```diff
@@ -0,0 +1,77 @@
+	return e.msg
+}
+
+func (e *timeoutError) Is(target error) bool {
+	return target == ErrFailureMsg
+}
```
What this says is "a timeout error is a ErrFailureMsg" and thus gets treated like one by `ExpectNoErrorWithOffset`.

pohly commented on test/e2e/framework/expect.go:
```diff
@@ -43,6 +47,18 @@ func ExpectNoError(err error, explain ...interface{}) {
 // (for example, for call chain f -> g -> ExpectNoErrorWithOffset(1, ...) error would be logged for "f").
 func ExpectNoErrorWithOffset(offset int, err error, explain ...interface{}) {
+	if err != nil &&
+		errors.Is(err, e2eerrors.ErrFailureMsg) {
```
That works for me. I didn't want to go that far because I wasn't sure whether it would reduce usability in other cases, but logging the error inside the test output should prevent that.

pohly committed:
e2e: refactor Ginkgo output testing

The helper code will be useful for testing other parts of the e2e framework.

pohly committed:
e2e: test output of e2epod wait

Both output during the test and the failure message are covered.

pohly committed:
e2e: make failure message for unexpected errors more readable

The failure message from Gomega was hard to read because explanation and error
text were separated by the error dump. In many cases, that error dump doesn't
add any relevant information.

Now the error is dumped first as info message (just in case that it is
relevant) and then a shorter failure message is created from explanation and
error text.

pohly commented:
I have rebased on top of master where unit tests in test/e2e get run.

The PR now uses the same kind of testing as test/e2e/framework/log_test.go, with the common code split out into a helper package. I am starting with testing of the current pod/wait.go without any modifications. Even if we don't agree on the other changes, then that is still worth merging because it provides unit testing for the changes from PR https://github.com/kubernetes/kubernetes/pull/109704

The next commits then make changes to failure reporting. The effect can be seen in the diff for the new pod/wait_test.go. I think these changes are useful, but it is debatable. For example, the "INFO: Unexpected error" log message becomes longer and fairly unreadable when including more text in the error message.

pohly commented:
/retest

tallclair commented on test/e2e/framework/pod/wait.go:
```diff
@@ -275,8 +275,8 @@ func WaitForPodCondition(c clientset.Interface, ns, podName, conditionDesc strin
-	if IsTimeout(err) {
-		e2elog.Logf("Timed out while waiting for pod %s to be %s. Last observed as: %s",
+	if IsTimeout(err) && lastPod != nil {
+		return TimeoutError("timed out while waiting for pod %s to be %s: last observed as:\n%s",
```
How does our test infrastructure aggregate failures? If anything is looking for identical failure messages, this will no longer match.

pohly replied:
Will no longer match because the dumped pod state will always be different, right?

The first line will still be the same, so searching for that should still work. I don't know whether that is going to be enough. Let me ask on #sig-testing.

pohly replied:
It looks like failure text differences are already handled:

https://storage.googleapis.com/k8s-triage/index.html?text=timed%20out%20while%20waiting

```
351 test failures (44 today) look like [link](https://storage.googleapis.com/k8s-triage/index.html?text=timed%20out%20while%20waiting#32d66067bb5fe4ce2e3e)[search github](https://github.com/search?type=Issues&q=org:kubernetes%2032d66067bb5fe4ce2e3e)[file bug](https://storage.googleapis.com/k8s-triage/index.html?text=timed%20out%20while%20waiting#)storage

test/e2e/storage/nfs_persistent_volume-disruptive.go:153
May  5 11:08:28.112: Unexpected error:
    <*errors.errorString | 0xc00654af50>: {
        s: "pod \"pvc-tester-8c6qg\" is not Running: timed out while waiting for pod disruptive-pv-142/pvc-tester-8c6qg to be running",
    }
    pod "pvc-tester-8c6qg" is not Running: timed out while waiting for pod disruptive-pv-142/pvc-tester-8c6qg to be running
occurred
test/e2e/storage/nfs_persistent_volume-disruptive.go:185
```

Note that pod name and namespace already vary in each failure instance.

pohly replied:
Ben explained [on Slack](https://kubernetes.slack.com/archives/C09QZ4DQB/p1652423870096959?thread_ts=1652423415.090849&cid=C09QZ4DQB) that this particular example works because varying strings like the 142 suffix get removed via regular expressions. Then remaining content must have <10% editing difference.

So including the pod could be a real problem. It could also help if the additional information let's the tool distinguish between different root causes for a failure, but I am a bit pessimistic about that.

Let's drop this part?

tallclair commented on test/e2e/framework/pod/wait.go:
```diff
@@ -275,8 +275,8 @@ func WaitForPodCondition(c clientset.Interface, ns, podName, conditionDesc strin
-		e2elog.Logf("Timed out while waiting for pod %s to be %s. Last observed as: %s",
+	if IsTimeout(err) && lastPod != nil {
+		return TimeoutError("timed out while waiting for pod %s to be %s: last observed as:\n%s",
 			podIdentifier(ns, podName), conditionDesc, format.Object(lastPod, 1))
```
(no action required, maybe a follow-up) I've already found these pod logs useful for debugging tests, but they are quite large and noisy. I'm thinking about using a yaml serializer instead, so that omitEmpty fields aren't (which would significantly cut down on the output size).

pohly replied:
I think that would be a useful enhancement for `gomega/format.Object`: skip fields that are marked as optional for JSON or YAML.

Otherwise we'll be stuck between a rock and a hard place when choosing a formatter: YAML serializers typically ignore unexported fields, which is bad for error types. output.Object prints those, but doesn't skip.

/cc @onsi

tallclair commented on test/e2e/framework/pod/wait_test.go:
```diff
@@ -0,0 +1,296 @@
+//
+//
+//
+//
```
ditto re: line number.
```suggestion
// This must be line 47.
```

tallclair commented on test/e2e/framework/pod/wait_test.go:
```diff
@@ -0,0 +1,296 @@
+})
+
+var (
+	podName      = "pending-pod"
+	podNamespace = "default"
```
nit: const. Also make the timeout a const.

pohly replied:
Done.

tallclair commented on test/e2e/framework/pod/wait_test.go:
```diff
@@ -0,0 +1,296 @@
+INFO: Pod "pending-pod": Phase="", Reason="", readiness=false. Elapsed: <elapsed>
+INFO: Unexpected error:
+    <*pod.timeoutError>: {
+        msg: "timed out while waiting for pod default/pending-pod to be running: last observed as:\n    <*v1.Pod>: {\n        TypeMeta: {Kind: \"\", APIVersion: \"\"},\n        ObjectMeta: {\n            Name: \"pending-pod\",\n            GenerateName: \"\",\n            Namespace: \"default\",\n            SelfLink: \"\",\n            UID: \"\",\n            ResourceVersion: \"\",\n             ...
```
:vomiting_face: 

This direct error output is always going to be a little gross, but one way you could clean this up is to include the pod directly in the TimeoutError struct, and just dump it when error is called. for example:

```go
type timeoutError struct {
	msg string
	lastObservedPod *v1.Pod  // or runtimeObject if we decide to generalize this
}

func (e *timeoutError) Error() string {
	if e.lastObservedPod == nil {
	  return e.msg
	} else {
	  return fmt.Sprintf("%s: last observed as: %s", e.msg, format.Object(e.lastObservedPod, 1))
	}
}

pohly replied:
Agreed. We could also make the timeout error generic and let it have a slice of objects that get stored as `interface{}`.

Would it make sense to split out the timeout error into its own package? There are other framework packages which implement polling, they may need the same timeout error eventually.

pohly replied:
My last commit now uses that approach, but without including the extra information in the error message. I am not adding it there because of the discussion around aggregation of failure messages - I am too worried that this won't work anymore.
I think it is a slight improvement because the overall output is shorter.

The timeout error is generic and could be moved as-is into a separate package, but I haven't done that yet.

If I did, should I use `test/e2e/framework/errors` or `test/e2e/framework/e2eerrors`? The latter avoids having to rename the package on each import to avoid clashes with the standard `errors` package, but it would be inconsistent because the other packages don't do that.

The double `ee` is also unfortunate. Perhaps `framework/e2etypes`?

pohly commented on test/e2e/framework/expect.go:
```diff
@@ -43,6 +46,18 @@ func ExpectNoError(err error, explain ...interface{}) {
+		}
+		Fail(prefix+err.Error(), 1+offset)
+	}
 	gomega.ExpectWithOffset(1+offset, err).NotTo(gomega.HaveOccurred(), explain...)
```
That's a leftover from the version where gomega was called for other, non-nil errors. Now that all non-nil errors are handled by our own code it can be removed.

pohly replied:
Done.

pohly committed:
e2e: include expected line number in output test

This makes it easier to see when comment lines must be added or removed.

tallclair commented on test/e2e/framework/expect.go:
```diff
@@ -43,7 +46,22 @@ func ExpectNoError(err error, explain ...interface{}) {
+			prefix = fmt.Sprintf("unexpected explain arguments, need format string: %v", explain)
+		}
+	}
+	Fail(prefix+err.Error(), 1+offset)
```
Consider adding a comment here as to why we're not using the default gomega assertions.

pohly replied:
Done.

tallclair commented on test/e2e/framework/log_test.go:
```diff
@@ -41,6 +35,12 @@ import (
+//
+//
+//
+// This must be line #43 (first line is #1).
```
What does `first line is #1` mean?

pohly replied:
That my editor is counting from 1. I'm not sure whether all of them do (first line might also be #0). Hmm, compilers and linters all print line numbers. I probably was too cautious here, there is an established convention. Let me remove it.

tallclair commented on test/e2e/framework/output/output.go:
```diff
@@ -0,0 +1,134 @@
+limitations under the License.
+*/
+
+package output
```
nit: should this be in a package `testing` to indicate it's for internal testing? I guess that's confusing since the whole thing is in the test tree...

pohly replied:
Or better, "internal"?

tallclair commented on test/e2e/framework/pod/wait_test.go:
```diff
@@ -0,0 +1,227 @@
+		output.TestResult{
+			Name: "[Top Level] pod not found",
+			Output: `INFO: Waiting up to 5s for pod "no-such-pod" in namespace "default" to be "running"
+INFO: Ignoring NotFound error while getting pod default/no-such-pod
+INFO: Ignoring NotFound error while getting pod default/no-such-pod
+INFO: Ignoring NotFound error while getting pod default/no-such-pod
+INFO: Ignoring NotFound error while getting pod default/no-such-pod
```
Is the number of attempts stable enough (especially on over subscribed CI infrastructure) that this won't be flaky?

pohly replied:
I wasn't sure about that myself. It polls every two seconds, so yes, I guess on a busy system it might skip one or more of these polls. I'm reducing those lines to one now.

tallclair commented on test/e2e/framework/expect.go:
```diff
@@ -43,7 +46,22 @@ func ExpectNoError(err error, explain ...interface{}) {
+
+	// Errors usually contain unexported fields. We have to use
+	// a formatter here which can print those.
+	Logf("Unexpected error:\n%s", format.Object(err, 1))
```
Consider including the explain text in this. Something like this?

```suggestion
	Logf("Unexpected error: %s\n%s", prefix, format.Object(err, 1))
```

pohly replied:
Makes sense, added.

pohly committed:
pods: more tolerant checking of poll tests

Under load we might skip some of the poll events, leading to less output. Now
we expect only at least one output line for polling.

tallclair reviewed (commented):
/lgtm
/approve

pohly commented on test/e2e/framework/log_test.go:
```diff
@@ -144,141 +132,56 @@ k8s.io/kubernetes/test/e2e/framework_test.glob..func1.5()
     }
-    an error with a long, useless description
-occurred
+FAIL: hard-coded error: an error with a long, useless description
```
This is an improvement. But after discussing it further on #sig-testing, I think I'll try to make it even better:

```
FAIL: <short failure message>
<details>
```

In this case:
```
FAIL: Unexpected error: hard-coded error: an error with a long, useless description
    <*errors.errorString>: {
        s: "an error with a long, useless description",
    }
```
